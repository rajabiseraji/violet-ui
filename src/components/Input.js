import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { mapToCssModules, warnOnce } from '../lib/utils';
import { CSSTransitionGroup } from 'react-transition-group';

const propTypes = {
  children: PropTypes.node,
  type: PropTypes.string,
  size: PropTypes.string,
  bsSize: PropTypes.string,
  valid: PropTypes.bool,
  tag: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  innerRef: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  plaintext: PropTypes.bool,
  addon: PropTypes.bool,
  className: PropTypes.string,
  cssModule: PropTypes.object
};

const defaultProps = {
  tag: 'p',
  type: 'text'
};

const FeedbackIcon = props => {
  let validIconClass = `fa fa-${props.valid
    ? 'check'
    : 'exclamation'} ${props.valid ? 'valid-icon' : 'invalid-icon'}`;
  if (props.valid === true || props.valid === false) {
    return <i key={props.valid + '-icon'} className={validIconClass} />;
  } else return null;
};

const InputWrapper = props => {
  return (
    <div
      key={props.innerRef + '-div'}
      className={classNames('inner-addon left-addon', props.className)}
    >
      {props.children}
    </div>
  );
};

class Input extends React.Component {
  render() {
    let {
      className,
      cssModule,
      type,
      bsSize,
      valid,
      tag,
      addon,
      plaintext,
      innerRef,
      ...attributes
    } = this.props;

    const checkInput = ['radio', 'checkbox'].indexOf(type) > -1;
    const isNotaNumber = new RegExp('\\D', 'g');

    const fileInput = type === 'file';
    const textareaInput = type === 'textarea';
    const selectInput = type === 'select';
    let Tag = selectInput || textareaInput ? type : 'input';

    let formControlClass = 'form-control';

    if (plaintext) {
      formControlClass = `${formControlClass}-plaintext`;
      Tag = tag;
    } else if (fileInput) {
      formControlClass = `${formControlClass}-file`;
    } else if (checkInput) {
      if (addon) {
        formControlClass = null;
      } else {
        formControlClass = 'form-check-input';
      }
    }

    if (attributes.size && isNotaNumber.test(attributes.size)) {
      warnOnce(
        'Please use the prop "bsSize" instead of the "size" to bootstrap\'s input sizing.'
      );
      bsSize = attributes.size;
      delete attributes.size;
    }

    const classes = mapToCssModules(
      classNames(
        className,
        valid === false && 'is-invalid',
        valid && 'is-valid',
        bsSize ? `form-control-${bsSize}` : false,
        formControlClass
      ),
      cssModule
    );

    if (Tag === 'input') {
      attributes.type = type;
    }

    if (Tag === 'input') {
      return (
        <InputWrapper
          valid={valid === undefined ? undefined : this.props.valid}
          innerRef={this.props.innerRef}
          className={this.props.className}
        >
          <CSSTransitionGroup
            transitionName="fade"
            transitionAppear={true}
            transitionAppearTimeout={500}
            transitionEnterTimeout={500}
            transitionLeaveTimeout={300}
            className={className}
          >
            <FeedbackIcon
              valid={this.props.valid}
              key={
                (this.props.valid === undefined ? '' : this.props.valid) +
                'feedback-icon'
              }
            />
          </CSSTransitionGroup>
          <input {...attributes} ref={innerRef} className={classes} />
        </InputWrapper>
      );
    } else {
      return <Tag {...attributes} ref={innerRef} className={classes} />;
    }
  }
}

Input.propTypes = propTypes;
Input.defaultProps = defaultProps;

export default Input;
