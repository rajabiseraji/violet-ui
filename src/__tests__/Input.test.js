import React from 'react';
import Input from '../components/Input';
import renderer from 'react-test-renderer';

/*
** This one is to make sure it later renders match the snapshot
*/

it('renders correctly', () => {
  const tree = renderer.create(<Input />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly in different sizes', () => {
  const sizes = ['sm', 'lg', 'md'];
  const component = renderer.create(<Input size="lg" />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  sizes.forEach(size => {
    tree.props.size = size;
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

it('Changes border color when valid changes in any size', () => {
  const sizes = ['sm', 'lg', 'md'];
  const component = renderer.create(<Input />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  sizes.forEach(size => {
    tree.props.size = size;
    tree.props.valid = true;
    // re-rendering
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    tree.props.valid = false;
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
