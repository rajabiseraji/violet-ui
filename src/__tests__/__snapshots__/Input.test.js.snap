// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Changes border color when valid changes in any size 1`] = `
<input
  className="form-control"
  type="text"
/>
`;

exports[`Changes border color when valid changes in any size 2`] = `
<input
  className="form-control"
  type="text"
/>
`;

exports[`Changes border color when valid changes in any size 3`] = `
<input
  className="form-control"
  type="text"
/>
`;

exports[`Changes border color when valid changes in any size 4`] = `
<input
  className="form-control"
  type="text"
/>
`;

exports[`Changes border color when valid changes in any size 5`] = `
<input
  className="form-control"
  type="text"
/>
`;

exports[`Changes border color when valid changes in any size 6`] = `
<input
  className="form-control"
  type="text"
/>
`;

exports[`Changes border color when valid changes in any size 7`] = `
<input
  className="form-control"
  type="text"
/>
`;

exports[`renders correctly 1`] = `
<input
  className="form-control"
  type="text"
/>
`;

exports[`renders correctly in different sizes 1`] = `
<input
  className="form-control-lg form-control"
  type="text"
/>
`;

exports[`renders correctly in different sizes 2`] = `
<input
  className="form-control-lg form-control"
  type="text"
/>
`;

exports[`renders correctly in different sizes 3`] = `
<input
  className="form-control-lg form-control"
  type="text"
/>
`;

exports[`renders correctly in different sizes 4`] = `
<input
  className="form-control-lg form-control"
  type="text"
/>
`;
